### PR TITLE
Np 44798 Expand AcademicChapter Anthology

### PR DESCRIPTION
**Pull request by monaullah**
An AcademicChapter has publicationContext type Anthology. Added expansion of Anthology with `mainTitle`, `contributors.identity.name` and `reference.publicationContext` if the expansion subject is an AcademicChapter.

### CODE REVIEW

File: expansion/src/main/java/no/unit/nva/expansion/model/ExpandedParentPublication.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    private static JsonLdOptions getJsonLdOptions() {
+        var jsonLdOptions = new JsonLdOptions();
+        jsonLdOptions.setOmitGraph(true);
+        jsonLdOptions.setOmitDefault(true);
+        jsonLdOptions.setUseNativeTypes(true);
+        jsonLdOptions.setPruneBlankNodeIdentifiers(true);
+        return jsonLdOptions;
+    }
```
Comment (brinxmat) [inline expansion/src/main/java/no/unit/nva/expansion/model/ExpandedParentPublication.java:73]: Move to helper class as these are now common across 3(?) classes:

```
package no.unit.nva.expansion;

import com.github.jsonldjava.core.JsonLdOptions;
import org.apache.jena.rdf.model.Model;
import org.apache.jena.riot.JsonLDWriteContext;
import org.apache.jena.riot.RDFFormat;
import org.apache.jena.riot.RDFWriter;

public class JsonLdDefaults {

    public static String frameJsonLd(Model model, String frame) {
        return RDFWriter.create()
                   .format(RDFFormat.JSONLD10_FRAME_PRETTY)
                   .context(getJsonLdWriteContext(frame))
                   .source(model)
                   .build()
                   .asString();
    }

    private static JsonLdOptions getJsonLdOptions() {
        var jsonLdOptions = new JsonLdOptions();
        jsonLdOptions.setOmitGraph(true);
        jsonLdOptions.setOmitDefault(true);
        jsonLdOptions.setUseNativeTypes(true);
        jsonLdOptions.setPruneBlankNodeIdentifiers(true);
        return jsonLdOptions;
    }
    



    private static JsonLDWriteContext getJsonLdWriteContext(String frame) {
        var context = new JsonLDWriteContext();
        context.setOptions(getJsonLdOptions());
        context.setFrame(frame);
        return context;
    }
}
```

File: expansion/src/main/java/no/unit/nva/expansion/model/ExpandedParentPublication.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    public String getExpandedParentPublication(URI publicationId) throws JsonProcessingException {
+        var publicationResponseBody = fetch(publicationId);
+        if (publicationResponseBody.isEmpty()) {
+            return EMPTY_STRING;
+        }
+        var model = createDefaultModel();
+        loadPublicationWithChannelDataIntoModel(publicationResponseBody.get(), model);
+        removePublicationTypeFromResource(publicationId, model);
+        return getFramedModelJson(model);
```
Comment (brinxmat) [inline expansion/src/main/java/no/unit/nva/expansion/model/ExpandedParentPublication.java:47]: Probably: 

```
    public String getExpandedParentPublication(URI publicationId) {
        return fetch(publicationId)
               .map(expandedPublication -> expandParent(publicationId, expandedPublication))
               .orElse(EMPTY_STRING);
    }

    private String expandParent(URI publicationId, String publicationResponseBody) {
        var model = createDefaultModel();
        loadPublicationWithChannelDataIntoModel(publicationResponseBody, model);
        removePublicationTypeFromResource(publicationId, model);
        return JsonLdDefaults.frameJsonLd(model, FRAME);
    }
```

and make the JSON-processing exception disappear with Try :)

File: expansion/src/main/java/no/unit/nva/expansion/model/ExpandedResource.java [hunk 3]
```diff
@@ -31,22 +32,24 @@
...
     // because is does not extend the Resource or Publication class,
     // but it contains its data as an inner Json Node.
-    public static final String ID_FIELD_NAME = "id";
+
     public static final String TYPE = "Publication";
-    public static final String JSON_LD_CONTEXT_FIELD = "@context";
+    private static final String ID_FIELD_NAME = "id";
+    private static final String JSON_LD_CONTEXT_FIELD = "@context";
+    private static final String CONTEXT_TYPE_ANTHOLOGY = "Anthology";
+    private static final String INSTANCE_TYPE_ACADEMIC_CHAPTER = "AcademicChapter";
```
Comment (brinxmat) [inline expansion/src/main/java/no/unit/nva/expansion/model/ExpandedResource.java:52]: It amuses me that we still use json-pointers unnecessarily — we should do something about that

File: expansion/src/main/java/no/unit/nva/expansion/model/IndexDocumentWrapperLinkedData.java [hunk 4]
```diff
@@ -23,26 +27,42 @@
...
 
-    public String toFramedJsonLd(JsonNode indexDocument) {
+    public String toFramedJsonLd(JsonNode indexDocument) throws JsonProcessingException {
         String frame = SearchIndexFrame.FRAME_SRC;
         List<InputStream> inputStreams = getInputStreams(indexDocument);
         return new FramedJsonGenerator(inputStreams, frame).getFramedJson();
     }
 
+    private static String extractIdField(JsonNode i) {
+        return i.at(ID_FIELD).asText();
```
Comment (brinxmat) [inline expansion/src/main/java/no/unit/nva/expansion/model/IndexDocumentWrapperLinkedData.java:47]: Any X that is a subject which is a Publication and is not also an object :D

File: expansion/src/main/java/no/unit/nva/expansion/model/IndexDocumentWrapperLinkedData.java [hunk 5]
```diff
@@ -23,26 +27,42 @@
...
+    }
+
+    private static Stream<JsonNode> extractParentAffiliationNodes(String affiliation) {
+        var json = attempt(() -> dtoObjectMapper.readTree(affiliation)).orElseThrow().at(PART_OF_FIELD);
+        return toStream(json);
+    }
```
Comment (brinxmat) [inline expansion/src/main/java/no/unit/nva/expansion/model/IndexDocumentWrapperLinkedData.java:53]: Again…another task to remove this code and use the model…

Comment (brinxmat) [reply expansion/src/main/java/no/unit/nva/expansion/model/IndexDocumentWrapperLinkedData.java:53]: But also:

```
    private static Stream<JsonNode> extractParentAffiliationNodes(String affiliation) {
        return attempt(() -> dtoObjectMapper.readTree(affiliation))
                   .map(IndexDocumentWrapperLinkedData::whateverThisIsDoing)
                   .map(IndexDocumentWrapperLinkedData::toStream)
                   .orElseThrow();
    }

    private static JsonNode whateverThisIsDoing(JsonNode node) {
        return node.at(PART_OF_FIELD);
    }
```

File: expansion/src/main/resources/parentPublicationFrame.json [hunk 6]
```diff
@@ -0,0 +1,202 @@
+{
```
Comment (brinxmat) [inline expansion/src/main/resources/parentPublicationFrame.json:1]: Can you run a diff on the now n+ frames to see what the differences between them are? I wonder if we really need them…all :)

File: expansion/src/test/java/no/unit/nva/publication/indexing/ExpandedResourceTest.java [hunk 7]
```diff
@@ -224,6 +242,36 @@ void shouldReturnIndexDocumentWithConfirmedJournalIdWhenPublicationIsPublishedIn
         assertThat(actualDocument.getPublicationContextUris(), containsInAnyOrder(expectedJournalId));
     }
 
+    @Test
+    void shouldReturnExpandedResourceWithAnthologyPublicationChannelUrisWhenPublicationIsAcademicChapter()
```
Comment (brinxmat) [inline expansion/src/test/java/no/unit/nva/publication/indexing/ExpandedResourceTest.java:246]: We should have a chat about other ways of testing this…

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: expansion/src/main/resources/parentPublicationFrame.json (hunk 6 above)
Comment (monaullah) [reply expansion/src/main/resources/parentPublicationFrame.json]: Merged frame used for nvi and frame used to create expanded publication for index

**Comment by monaullah** [comment]
Live code review with @brinxmat, approved

**Pull request merged**